source/core/icosahedron.ts

```diff
@@ -0,0 +1,212 @@
+        /* Create cache to avoid duplicating vertices. */
+        const cache = new Map<number, number>();
+
+        /* Execute specified number of refinement levels. */
+        for (let i = 0; i < levels; i++) {
+            const size = faces.length;
+
+            /* Refine each face of the current level. */
+            for (let f = 0; f < size; f++) {
+                /* Get face to refine */
+                const face = faces[f];
+
+                /* Get vertices of the current face. */
+                const a = face[0];
+                const b = face[1];
+                const c = face[2];
+
+                /* Calculate center points of each edge. */
+                const ab = this.split(a, b, vertices, cache);
+                const bc = this.split(b, c, vertices, cache);
+                const ca = this.split(c, a, vertices, cache);
+
+                /* Split triangle into 4. */
+                faces[f] = [ab, bc, ca];
+                faces.push([a, ab, ca]);
+                faces.push([b, bc, ab]);
+                faces.push([c, ca, bc]);
+            }
+        }
+    }
+
+    /**
+     * Splits an edge and returns the new vertex that is the center of the edge.
+     * @param a - Index of first vertex.
+     * @param b - Index of second vertex.
+     * @param vertices - List of vertices that is modified.
+     * @param cache - Cache to identify newly created vertices.
+     */
+    protected split(a: number, b: number, vertices: Array<vec3>, cache: Map<number, number>): number {
+        /* Create hash to identify the new vertex. */
+        const aSmaller: boolean = (a < b);
+        const smaller = aSmaller ? a : b;
+        const greater = aSmaller ? b : a;
+        const hash = (smaller << 16) + greater;
+
+        /* Check if a vertex between these two has already been created. */
+        if (cache.has(hash)) {
+            return cache.get(hash)!;
+        }
+
+        /* Create vertex between these two, normalize it to form a point on the sphere. */
+        const pos = vec3.add(v3(), vertices[a], vertices[b]);
+        vec3.normalize(pos, pos);
+        vertices.push(pos);
+
+        /* Put new vertex into the cache. */
+        const index = vertices.length - 1;
+        cache.set(hash, index);
+
+        /* Return index of the new vertex. */
+        return index;
+    }
+
+    /**
+     * Generates the geometry of the icosahedron.
+     * @param levels - Number of levels of refinement.
+     */
+    generateGeometry(levels: number): void {
+        /* Create base geometry. */
+        this._vertices = this.baseVertices();
+        this._faces = this.baseFaces();
+
+        /* Refine geometry. */
+        if (levels > 0) {
+            this.refine(this._vertices, this._faces, Math.min(Math.max(levels, 0), 8));
+        }
+    }
+
+    /**
+     * Generates texture coordinates for the icosahedron.
+     */
+    generateTextureCoordinates(): void {
```

Comment:
per flag in initialize `... (..., generateTexCoords: boolean = true);` - and make function protected or include within `initialize`